### PR TITLE
RFE: Templating Support for Definition File

### LINKED ISSUE #963: RFE: Templating Support for Definition File
Description
-----------
If we can pass values at build time to replace variables' value inside definition file, before actually process definition file to build image, that brings flexibility to how we interact with definition file. I called this feature, "Templateing".

Usually OSS projects distribute not only source code but its binaries and also container images when they release these days. In the CI/CD era, creating binaries is written in the form of scripts that covers wide range of OS and variety of dependencies. That "recipe" can switch OS versions, library versions etc. Widely known examples in HPC are Spack recipes.

If Apptainer definition file support "Templating" feature:

+ Less integration effort for OSS maintainers: They can reuse existing "recipe" scripts when they try to add Apptainer definition file to their release artifact. 
+ It allows users to alter base image and/or version (This is not supported in Apptainer yet.)
+ It allows users to alter script "switch" (This is kind of supported through APPTAINERENV_VARS but this is not support every section in definition file.)
+ It allows users to alter variables at every section in consistent way (This is not supported in Apptainer yet.)

This is inspired by Dockerfile `ARG` and `--build-arg` (`--build-arg-file`) option for build command.

UI Changes
------------------------
Add `--build-arg` and `--build-arg-file` options to `build` command

`--build-arg`

```
apptainer build --build-arg OS_VER=8.7 --build-arg APP_VERSION=2206 app.sif app.def
```

`--build-arg-file`

```
apptainer build --build-arg-file build-args app.sif app.def
```

`build-args` file

```
OS_VER=8.7
APP_VERSION=2206
```

Internal Behavioural Changes
------------------------

When `build-arg` or `build-arg-file` option comes in:
1. process build-args first
2. fill template to create actual definition file
3. process definition file

How definition file looks like:
-------------------------------

```
Bootstrap: docker
From: rockylinux:${OS_VER}

%setup
    touch /${OS_VER}
    touch ${APPTAINER_ROOTFS}/file2

%files
    /script-${OS_VER}.sh
    /dir-${OS_VER} /opt

%environment
    export OS_VER=${OS_VER}

%post
    /script-${OS_VER}.sh ${APP_VER}

%runscript
    echo "Container was created $NOW"
    echo "Arguments received: $*"
    exec echo "$@"

%startscript
    nc -lp ${LISTEN_PORT:-'8080'}

%test
    grep -q NAME=\"Rocky\" /etc/os-release
    if [ $? -eq 0 ]; then
        echo "Container base is Rocky Linux as expected."
    else
        echo "Container base is not Rocky Linux."
        exit 1
    fi

%labels
    Author ${AUTHER:-}
    Version ${VERSION:-}

%help
    This is a ${DEMO:-} for templating definition file
```

I am please to hear any comments and ideas from community here.

### PR DESCRIPTION
**Pull request by JasonYangShadow**
## Description of the Pull Request (PR):

Add template support for build procedure. 

### This fixes or addresses the following GitHub issues:

 - Fixes #963

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

### CODE REVIEW

**Comment by JasonYangShadow** [comment]
Feature description starts from here
https://github.com/apptainer/apptainer/issues/963#issuecomment-1527130389

File: CHANGELOG.md [hunk 1]
```diff
@@ -63,6 +63,7 @@ For older changes see the [archived Singularity change log](https://github.com/a
 - The `instance start` command now accepts an optional `--app <name>` argument which
   invokes start script within the `%appstart <name>` section in the definition file.
   The `instance stop` command still only requires the instance name.
+- Templating Support for Definition File.
```
Comment (DrDaveD) [inline CHANGELOG.md:66]: Please expand this to mention the new `--build-arg` and `--build-arg-file` options for replacing variables in definition files that are surrounded by a matching pair of double curly brackets `{{ ... }}`.

File: cmd/internal/cli/build.go [hunk 2]
```diff
@@ -285,6 +287,22 @@ var buildRemoteFlag = cmdline.Flag{
 	Hidden:       true,
 }
 
+var buildVarArgsFlag = cmdline.Flag{
+	ID:           "buildVarArgsFlag",
+	Value:        &buildArgs.buildVarArgs,
+	DefaultValue: []string{},
+	Name:         "build-arg",
+	Usage:        "--build-arg is used for passing variables to build procedure",
```
Comment (DrDaveD) [inline cmd/internal/cli/build.go:295]: Don't include the variable name in the Usage.  Use somethng like "defines variable=value to replace '{{variable}}' with value in build definition files".

File: cmd/internal/cli/build.go [hunk 3]
```diff
@@ -285,6 +287,22 @@ var buildRemoteFlag = cmdline.Flag{
...
+}
+
+var buildVarArgFileFlag = cmdline.Flag{
+	ID:           "buildVarArgFileFlag",
+	Value:        &buildArgs.buildVarArgFile,
+	DefaultValue: "",
+	Name:         "build-arg-file",
+	Usage:        "--build-arg-file is used for passing variables file to build procedure",
```
Comment (DrDaveD) [inline cmd/internal/cli/build.go:303]: Use "specifies a file containing variable=value lines to replace '{{variable}}' with value in build definition files"

File: cmd/internal/cli/build_linux.go [hunk 4]
```diff
@@ -468,3 +477,146 @@ func getEncryptionMaterial(cmd *cobra.Command) (*cryptkey.KeyInfo, error) {
...
+	if val == "" {
+		return "", "", fmt.Errorf("value field is missing in text: %s", text)
+	}
+	return key, val, nil
+}
+
+var errNoChange = errors.New("text no change")
+
+func replaceVar(text []byte, argsMap map[string]string, defaultMap map[string]string) ([]byte, error) {
+	r := regexp.MustCompile(`{{\s*(\w*)\s*}}`)
```
Comment (DrDaveD) [inline cmd/internal/cli/build_linux.go:565]: I assume that `\w*` means any number of non-space characters, including zero.  I would think you want to make sure there's at least one non-space character.

File: cmd/internal/cli/build_linux.go [hunk 5]
```diff
@@ -468,3 +477,146 @@ func getEncryptionMaterial(cmd *cobra.Command) (*cryptkey.KeyInfo, error) {
...
+		return "", "", fmt.Errorf("key field is missing in text: %s", text)
+	}
+	val := strings.TrimSpace(matches[1])
+	if val == "" {
+		return "", "", fmt.Errorf("value field is missing in text: %s", text)
+	}
+	return key, val, nil
+}
+
+var errNoChange = errors.New("text no change")
```
Comment (DrDaveD) [inline cmd/internal/cli/build_linux.go:562]: How about "no change to text" instead.

**Review by DrDaveD — CHANGES_REQUESTED** [review]

**Comment by panda1100** [comment]
@JasonYangShadow Thank you Jason!!

I tried the following with regular user 

```
apptainer build --build-arg-file=args.list 963-from-file.sif 963.def
```

This fails due to `/root/APPTAINER_PR/test.sh: No such file or directory`, I used `~/` inside `args.list`

The contents of `args.list`:

```
OS=rockylinux
VERSION=9
SCRIPT_PATH=~/APPTAINER_PR/test.sh
```

`963.def`

```
Bootstrap: docker
From: {{ OS }}:{{ VERSION }}

%default
    AUTHOR=ysenda
    VERSION=1
    DEMO=awesome demo
    TMP_FILE=tmpfile

%setup
    touch ${APPTAINER_ROOTFS}/{{ TMP_FILE }}

%files
    {{ SCRIPT_PATH }}

%environment
    export AUTHOR={{ AUTHOR }}
    export VERSION={{ VERSION }}

%post
    NOW=`date`
    {{ SCRIPT_PATH }}

%runscript
    echo "Container was created $NOW"

%startscript
    wget --help

%labels
    Author {{ AUTHOR }}
    Version {{ VERSION }}

%help
    This is a {{ DEMO }} for templating definition file
```

The following works perfect as expected.

```
apptainer build --build-arg OS=rockylinux --build-arg VERSION=9 --build-arg SCRIPT_PATH=~/APPTAINER_PR/test.sh  963.sif 963.def
```

a bit of context of my environment

```
INFO:    User not listed in /etc/subuid, trying root-mapped namespace
INFO:    The %post section will be run under fakeroot
```

**Comment by JasonYangShadow** [comment]
> @JasonYangShadow Thank you Jason!!
> 
> I tried the following with regular user
> 
> ```
> apptainer build --build-arg-file=args.list 963-from-file.sif 963.def
> ```
> 
> This fails due to `/root/APPTAINER_PR/test.sh: No such file or directory`, I used `~/` inside `args.list`
> 
> The contents of `args.list`:
> 
> ```
> OS=rockylinux
> VERSION=9
> SCRIPT_PATH=~/APPTAINER_PR/test.sh
> ```
> 
> `963.def`
> 
> ```
> Bootstrap: docker
> From: {{ OS }}:{{ VERSION }}
> 
> %default
>     AUTHOR=ysenda
>     VERSION=1
>     DEMO=awesome demo
>     TMP_FILE=tmpfile
> 
> %setup
>     touch ${APPTAINER_ROOTFS}/{{ TMP_FILE }}
> 
> %files
>     {{ SCRIPT_PATH }}
> 
> %environment
>     export AUTHOR={{ AUTHOR }}
>     export VERSION={{ VERSION }}
> 
> %post
>     NOW=`date`
>     {{ SCRIPT_PATH }}
> 
> %runscript
>     echo "Container was created $NOW"
> 
> %startscript
>     wget --help
> 
> %labels
>     Author {{ AUTHOR }}
>     Version {{ VERSION }}
> 
> %help
>     This is a {{ DEMO }} for templating definition file
> ```
> 
> The following works perfect as expected.
> 
> ```
> apptainer build --build-arg OS=rockylinux --build-arg VERSION=9 --build-arg SCRIPT_PATH=~/APPTAINER_PR/test.sh  963.sif 963.def
> ```
> 
> a bit of context of my environment
> 
> ```
> INFO:    User not listed in /etc/subuid, trying root-mapped namespace
> INFO:    The %post section will be run under fakeroot
> ```

As we discussed separately, using absolute path is good approach to prevent such issues. `~` can be interpreted into different value in different sections of the definition file.

File: CHANGELOG.md (hunk 1 above)
Comment (JasonYangShadow) [reply CHANGELOG.md]: thanks! resolved in the latest commit.
https://github.com/apptainer/apptainer/pull/1344/commits/1842bc374a7db1c7a72fb8fa11d811d7de4e2ff4

File: cmd/internal/cli/build.go (hunk 2 above)
Comment (JasonYangShadow) [reply cmd/internal/cli/build.go]: thanks! resolved in the latest commit.
https://github.com/apptainer/apptainer/pull/1344/commits/1842bc374a7db1c7a72fb8fa11d811d7de4e2ff4

File: cmd/internal/cli/build.go (hunk 3 above)
Comment (JasonYangShadow) [reply cmd/internal/cli/build.go]: thanks! resolved in the latest commit.
https://github.com/apptainer/apptainer/pull/1344/commits/1842bc374a7db1c7a72fb8fa11d811d7de4e2ff4

File: cmd/internal/cli/build_linux.go (hunk 4 above)
Comment (JasonYangShadow) [reply cmd/internal/cli/build_linux.go]: thanks! resolved in the latest commit.
https://github.com/apptainer/apptainer/pull/1344/commits/1842bc374a7db1c7a72fb8fa11d811d7de4e2ff4

File: cmd/internal/cli/build_linux.go (hunk 5 above)
Comment (JasonYangShadow) [reply cmd/internal/cli/build_linux.go]: thanks! resolved in the latest commit.
https://github.com/apptainer/apptainer/pull/1344/commits/1842bc374a7db1c7a72fb8fa11d811d7de4e2ff4

**Comment by panda1100** [comment]
@JasonYangShadow I tested with absolute path and that absolutely works :smile_cat:

**Comment by panda1100** [comment]
This is totally expected behavior but I'll share anyway.

Building  template using Apptainer v1.1.8 will show you the following.

```
$ apptainer build 963-v118.sif 963.def
FATAL:   Unable to build from 963.def: while parsing definition: 963.def: invalid section(s) specified: default
```

If we have SIF file, we can extract definition file with variables filled at the build time 

```
$ apptainer sif list 963.sif
------------------------------------------------------------------------------
ID   |GROUP   |LINK    |SIF POSITION (start-end)  |TYPE
------------------------------------------------------------------------------
1    |1       |NONE    |32176-32702               |Def.FILE
2    |1       |NONE    |32702-34831               |JSON.Generic
3    |1       |NONE    |34831-34928               |JSON.Generic
4    |1       |NONE    |36864-62242816            |FS (Squashfs/*System/amd64)
```

```
$ apptainer sif dump 1 963.sif > filled_template_963.def
```

```
$ cat filled_template_963.def
bootstrap: docker
from: rockylinux:9

%labels
	Author ysenda
	Version 9

%files
	/home/ysenda/APPTAINER_PR/test.sh	

%help
    This is a awesome demo for templating definition file


%environment
    export AUTHOR=ysenda
    export VERSION=9



%runscript
    echo "Container was created $NOW"



%startscript
    wget --help



%setup
    touch ${APPTAINER_ROOTFS}/tmpfile



%post
    NOW=`date`
    /home/ysenda/APPTAINER_PR/test.sh



%default
    AUTHOR=ysenda
    VERSION=1
    DEMO=awesome demo
    TMP_FILE=tmpfile
```

**Comment by panda1100** [comment]
Hi @olesenm Mark-san, This is PR for Apptainer template support. Please feel free to drop some questions and/or comments, suggestions here. Thank you!

File: CHANGELOG.md [hunk 6]
```diff
@@ -63,6 +63,10 @@ For older changes see the [archived Singularity change log](https://github.com/a
 - The `instance start` command now accepts an optional `--app <name>` argument which
   invokes start script within the `%appstart <name>` section in the definition file.
   The `instance stop` command still only requires the instance name.
+- Templating support for definition file. Users can define variables in definition
+  file via a matching pair of double curly brackets `{{ variable }}`, defined variables
+  will be replaced later either by `variable=value` entry defined in `%default` section
+  in definition file or through new cli options `--build-arg` and `--build-arg-file`.
```
Comment (DrDaveD) [inline CHANGELOG.md:69]: A little cleanup of the English:
```
- Templating support for definition files. Users can now define variables in definition
  files via a matching pair of double curly brackets. Variables of the form `{{ variable }}` 
  will be replaced by a value defined either by a `variable=value` entry in the `%default`
  section of the definition file or through new build options `--build-arg` or `--build-arg-file`.
```

**Review by DrDaveD — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 6 above)
Comment (JasonYangShadow) [reply CHANGELOG.md]: resolved and squashed in this commit
https://github.com/apptainer/apptainer/pull/1344/commits/377b9e8a8bd299caef150607c1463085acd871db

**Comment by olesenm** [comment]
> Hi @olesenm Mark-san, This is PR for Apptainer template support. Please feel free to drop some questions and/or comments, suggestions here. Thank you!

Unfortunately I'm short of time for the next while (holidays, release etc) so can't take a look at all - sorry

**Comment by panda1100** [comment]
Thank you @olesenm , I look forward to new release! and no worries, I just wanted to let you know this feature has been made some progress.

**Pull request closed without merging**